File: frontend_mobile/lib/data/models/notification/notification_dto.dart
import 'package:json_annotation/json_annotation.dart';

part 'notification_dto.g.dart';

@JsonSerializable()
class NotificationDto {
  final String publicId;
  final String type; // RECIPE_COOKED, RECIPE_VARIATION
  final String title;
  final String body;
  final String? recipePublicId;
  final String? logPostPublicId;
  final String? senderUsername;
  final String? senderProfileImageUrl;
  final bool isRead;
  final DateTime createdAt;
  final Map<String, dynamic>? data;

  NotificationDto({
    required this.publicId,
    required this.type,
    required this.title,
    required this.body,
    this.recipePublicId,
    this.logPostPublicId,
    this.senderUsername,
    this.senderProfileImageUrl,
    required this.isRead,
    required this.createdAt,
    this.data,
  });

  factory NotificationDto.fromJson(Map<String, dynamic> json) =>
      _$NotificationDtoFromJson(json);

  Map<String, dynamic> toJson() => _$NotificationDtoToJson(this);

  NotificationDto copyWith({bool? isRead}) {
    return NotificationDto(
      publicId: publicId,
      type: type,
      title: title,
      body: body,
      recipePublicId: recipePublicId,
      logPostPublicId: logPostPublicId,
      senderUsername: senderUsername,
      senderProfileImageUrl: senderProfileImageUrl,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt,
      data: data,
    );
  }
}

File: backend/src/main/java/com/cookstemma/cookstemma/dto/notification/NotificationDto.java
package com.cookstemma.cookstemma.dto.notification;

import com.cookstemma.cookstemma.domain.entity.notification.Notification;
import com.cookstemma.cookstemma.domain.enums.NotificationType;

import java.time.Instant;
import java.util.Map;
import java.util.UUID;

public record NotificationDto(
    UUID publicId,
    NotificationType type,
    String title,
    String body,
    UUID recipePublicId,
    UUID logPostPublicId,
    String senderUsername,
    String senderProfileImageUrl,
    Boolean isRead,
    Instant createdAt,
    Map<String, Object> data
) {
    public static NotificationDto from(Notification n) {
        return new NotificationDto(
            n.getPublicId(),
            n.getType(),
            n.getTitle(),
            n.getBody(),
            n.getRecipe() != null ? n.getRecipe().getPublicId() : null,
            n.getLogPost() != null ? n.getLogPost().getPublicId() : null,
            n.getSender() != null ? n.getSender().getUsername() : null,
            n.getSender() != null ? n.getSender().getProfileImageUrl() : null,
            n.getIsRead(),
            n.getCreatedAt(),
            n.getData()
        );
    }
}
